File: Weekly_Contests/Biweekly_99/Q4.py
class Solution:
    def rootCount(self, edges: List[List[int]], guesses: List[List[int]], k: int) -> int:
        # literally try every possible node as root
        # count the correct guess by using dfs and memoization since it's tree
        # from a given node 1 to node 2, there's only one path
        graph = collections.defaultdict(list)
        for u, v in edges:
            graph[u].append(v)
            graph[v].append(u)
        
        seen = set()
        for u, v in guesses:
            seen.add((u, v))
        
        mem = {}

        def correct_guess(i, prev):
            if (i, prev) in mem:
                return mem[(i, prev)]
            mem[(i, prev)] = 0
            for nxt in graph[i]:
                if nxt == prev:
                    continue
                if (i, nxt) in seen:
                    mem[(i, prev)] += 1
                mem[(i, prev)] += correct_guess(nxt, i)
            return mem[(i, prev)]

        ans = 0
        for i in range(len(edges) + 1):
            if correct_guess(i, -1) >= k:
                ans += 1
        return ans

File: Weekly_Contests/Biweekly_99/Q4.cpp
class Solution {
public:
    int correct_guess(int i, int prev, unordered_map<int, unordered_map<int, int>>& mem, unordered_map<int, unordered_map<int, int>>& seen, unordered_map<int, vector<int>>& graph){
        if (mem.count(i) && mem[i].count(prev)){
            return mem[i][prev];
        }
        mem[i][prev] = 0;
        for (int nxt: graph[i]){
            //cout << i << " " << nxt << endl;
            if (nxt == prev){
                continue;
            }
            if (seen.count(i) && seen[i].count(nxt)){
                //cout << i << " " << prev << endl;
                mem[i][prev]++;
            }
            mem[i][prev] += correct_guess(nxt, i, mem, seen, graph);
        }
        return mem[i][prev];
    }

    int rootCount(vector<vector<int>>& edges, vector<vector<int>>& guesses, int k) {
        int n = edges.size() + 1;
        unordered_map<int, vector<int>> graph;
        for (vector<int> e : edges){
            int u = e[0], v = e[1];
            graph[u].push_back(v);
            graph[v].push_back(u);
        }
        unordered_map<int, unordered_map<int, int>> seen;
        for (vector<int> g : guesses){
            int u = g[0], v = g[1];
            seen[u][v] = 1;
        }
        unordered_map<int, unordered_map<int, int>> mem;

        int ans = 0;
        for (int i = 0; i < n; i++){
            int correct = correct_guess(i, -1, mem, seen, graph);
            //cout << i << " " << correct << endl;
            if (correct >= k){
                ans++;
            }
        }
        return ans;
    }
};